### ISSUE TITLE
Vulnerability roundup 91: busybox-1.31.1: 1 advisory [8.1]

### ISSUE THREAD

**Issue by ckauhaus** [NixOS/nixpkgs#94722]
[search](https://search.nix.gsc.io/?q=busybox&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=busybox+in%3Apath&type=Code)

* [ ] [CVE-2018-1000500](https://nvd.nist.gov/vuln/detail/CVE-2018-1000500) CVSSv3=8.1 (nixos-20.03, nixos-unstable)

Scanned versions: nixos-20.03: 977000f149b; nixos-unstable: 840c782d507. May contain false positives.

Cc @TethysSvensson

**Comment by TethysSvensson** [comment]
@Mic92

The nixpkgs in both 20.03 and current master is 1.31.1. The [patch](https://git.busybox.net/busybox/commit/?id=45fa3f18adf57ef9d743038743d9c90573aeeb91) is not included until busybox 1.32.0, which is marked as an unstable release.

As far as I can tell, we do not by default use busybox in a vulnerable configuration, however somebody could easily compile a vulnerable version by mistake by modifying `extraConfig`.

What do you think? Should we backport the patch to 1.31.1?

**Comment by TethysSvensson** [comment]
I tested this by running the example from [the ubuntu bug report](https://bugs.launchpad.net/ubuntu/+source/busybox/+bug/1879533):

```
~> nix-shell -p busybox

[nix-shell:~]$ busybox wget https://untrusted-root.badssl.com/
Connecting to untrusted-root.badssl.com (104.154.89.105:443)
wget: note: TLS certificate validation not implemented
saving to 'index.html'
index.html           100% |*************************************************************|   600  0:00:00 ETA
'index.html' saved
```

As shown, we do get a warning in the default configuration.

**Comment by Mic92** [comment]
Is this the complete fix? https://git.busybox.net/busybox/commit/?id=45fa3f18adf57ef9d743038743d9c90573aeeb91 I don't see how they enable TLS validation there.

**Comment by TethysSvensson** [comment]
I think they enable it with the `-verify_return_error` option.

**Comment by Mic92** [comment]
> I think they enable it with the `-verify_return_error` option.

I missed the `!` in the if condition. Now it makes more sense. We can cherry-pick this commit.

**Comment by TethysSvensson** [comment]
On closer inspection, we are vulnerable even in the default configuration. Busybox will first try to connect using the `openssl` binary, but if it is unavailable it will use internal tls code instead. A warning is only printed if the internal tls code is used, while the `openssl` binary will accept invalid certificates without any warnings. How to reproduce:

```
~> nix-shell -p busybox openssl

[nix-shell:~]$ busybox wget https://untrusted-root.badssl.com/
Connecting to untrusted-root.badssl.com (104.154.89.105:443)
saving to 'index.html'
index.html           100% |****************************************************************************************|   600  0:00:00 ETA
'index.html' saved
```

I will try to crate a PR with the patch.

**Comment by Mic92** [comment]
Fixed in master and 20.03.
